keps/sig-cluster-lifecycle/kubeadm/20190424-kubeadm-for-windows.md

```diff
@@ -171,76 +183,53 @@ Windows related adjustments to default paths might be required.
-
-The proposed Windows wrapper script that executes kubeadm will also manage the restart of the kube-proxy Windows service.
-
-Long term and ideally, kube-proxy should be run as a DaemonSet on Windows.
+On Linux, kube-proxy is deployed as a DaemonSet in the kubeadm init phase. We will supply a DaemonSet that will run on Windows and use the wins API to launch kube-proxy.
 
 #### CNI plugin deployment
 
-On Linux, CNI plugins are deployed via kubectl and run as a DaemonSet. However, on Windows, CNI plugins need to run on the node, and cannot run in containers (again because Windows does not currently support privileged containers). Azure-CNI, win-bridge (compatible with kubenet), and Flannel all need the binary and config stored on the node.
+On Linux, CNI plugins are deployed via kubectl and run as a DaemonSet. We will
+supply DaemonSets that will allow users to run Flannel configured in either
+L2Bridge/Host-Gateway or VXLAN/Overlay mode. These ideally can be upstreamed into the Flannel
+project.
 
-This proposal plans for FlannelD as the default option. Currently, FlannelD has to be started before the kube-proxy Windows service is started. FlannelD creates an HNS network on the Windows host, and kube-proxy will crash if it cannot find the network. This should be fixed in the scope of this project so that kube-proxy will wait until the network comes up. Therefore, kube proxy can be started at any time. 
+If users wish to use a different plugin they can create a DaemonSet to be applied in place of the Flannel DaemonSet.
 
-However, if FlannelD is deployed in VXLAN (Overlay) mode, then we need to rewrite the KubeProxyConfiguration with the correct Overlay specific values, and kube-proxy will need to read this config again. This is not true for Host-Gateway (L2Bridge) mode. The script will have a flag that allows users to choose between the two networking modes.
+### Risks and Mitigations
 
-If the users wish to use a different plugin they will technically opt-out of the supported setup for kubeadm based Windows worker nodes in the Alpha release.
+**Risk**: Wins proxy introduces new security vector
 
-Long term, any CNI plugin should be supported for kubeadm based Windows worker nodes.
+The same functionality that allows us to now run privileged DaemonSets on Windows could be used maliciously to perform
+unwanted behavior on Windows nodes. This brings to Windows problems that already exist on Linux and now require the same
+mitigations, namely Pod Security Policies (PSP).
 
-### Risks and Mitigations
+*Mitigation*: Access to the wins named pipe can be restricted using a PSP that either disables
+`hostPath` volume mounts or restricts the paths that can be mounted. A sample PSP will be provided.
```

Comment:
I'm not wild about depending on PSP to enforce this behavior. Here are 2 alternative ideas:

1. Custom admission controller for dealing with windows pods. Only allow the `wins` pipe to be mounted (applies to the whole directory tree - does windows support hard linking?) if the requesting container is running as `privileged`.
2. Put an authenticating proxy in front of wins. Require a service account token to make privileged requests, and check against the podspec whether a request with that privilege level should be allowed. I'm not sure I understand how windows containers / wins work well enough to know whether this approach would be viable.
3. (A weaker version of 2) Use the authenticating proxy, but just check for the presence of a static secret. The secret would be added to a kubernetes secret, and mounted into containers that need to access wins. Perhaps rancher would consider an upstream patch to build this capability in?

Reply:
> I'm not wild about depending on PSP to enforce this behavior.

i concur.

would defer to @benmoss and @PatrickLang to the Windows specific questions.

1 seems fine to me, but if understand this correctly it requires that the controller is included in core k8s. AFAIK, wins currently lacks an authorization / authentication model so maybe this is the way forward as long as the project is OK with that.

Reply:
Thanks @tallclair, @neolit123

Reply:
Thanks @tallclair @neolit123 

We arrived at using PSPs by following the pattern for privileged containers in Linux as privileged containers were kind of similar to what was being offered by wins. Can you elaborate on your reluctance to use PSPs? 
- Are PSPs insufficient in protecting against malicious pods mounting that wins pipe?
- Is it the idea of relying on an operator to apply PSPs and not having things secured out of the box that's hard to accept?